source/crud/bulk-write.md

```diff
@@ -0,0 +1,724 @@
+
+When auto-encryption is enabled, drivers MUST NOT provide the `ops` and `nsInfo` fields as document
+sequences and MUST limit the size of the command according to the auto-encryption size limits
+defined in the
+[Client Side Encryption Specification](../client-side-encryption/client-side-encryption.rst).
+
+#### Unencrypted bulk writes with document sequences
+
+When `ops` and `nsInfo` are provided as document sequences, drivers MUST ensure that the total size
+of the `OP_MSG` built for the `bulkWrite` command does not exceed `maxMessageSizeBytes`. Some
+drivers may perform batch-splitting prior to constructing the full `OP_MSG` to be sent to the
+server. In this case, drivers MAY use `maxMessageSizeBytes - 16,384` as the upper bound for the
+combined number of bytes in `ops` and `nsInfo`. 16KiB is subtracted to accommodate for the bytes in
+the rest of the message.
+
+#### Unencrypted bulk writes without document sequences
+
+When `ops` and `nsInfo` are provided within the `bulkWrite` command document and auto-encryption is
+not enabled, drivers MUST ensure that the total size of the `bulkWrite` command document does not
+exceed `maxBsonObjectSize`.
+
+## Handling the `bulkWrite` Server Response
+
+The server's response to `bulkWrite` has the following format:
+
+```json
+{
+    "ok": <0 | 1>,
+    "cursor": {
+        "id": <Int64>,
+        "firstBatch": <Array>,
+        "ns": <String>
+    },
+    "nErrors": <Int32>,
+    "nInserted": <Int32>,
+    "nUpserted": <Int32>,
+    "nMatched": <Int32>,
+    "nModified": <Int32>,
+    "nDeleted": <Int32>,
+    ...additional command-agnostic fields
+}
+```
+
+If any operations were successful (i.e. `nErrors` is less than the number of operations that were
+sent), drivers MUST record the summary count fields in a `BulkWriteResult` to be returned to the
+user or embedded in a `BulkWriteException`. Drivers MUST NOT populate the `partialResult` field in
+`BulkWriteException` if no operations were successful.
+
+Drivers MUST attempt to consume the contents of the cursor returned in the server's `bulkWrite`
+response before returning to the user. This is required regardless of whether the user requested
+verbose or summary results, as the results cursor always contains any write errors that occurred.
+If the cursor contains a nonzero cursor ID, drivers MUST perform `getMore`s until the cursor has
+been exhausted. Drivers MUST use the same session used for the `bulkWrite` command for each
+`getMore` call.
+
+The documents in the results cursor have the following format:
+
+```json
+{
+    "ok": <0 | 1>,
+    "code": Optional<Int32>,
+    "errmsg": Optional<String>,
+    "n": <Int32>,
+    "nMatched": <Int32>,
+    "nModified": <Int32>,
+    "upsertedId": Optional<ObjectId>
+}
+```
+
+Note that the responses do not contain information about the type of operation that was performed.
+Drivers may need to maintain the user's list of write models to infer which type of result should
+be recorded.
```

Comment:
How do drivers correlate cursor documents with original operations? Particularly, in an errorsOnly scenario, I assume the cursor would only report errorred operations. I don't see any field in the response that indicates the index of a write operation.